---
layout:     post
title:      oncoplot绘制
date:       2024-03-01
author:     champeil
catalog: true
tags:
    - wes
    - wgs
    - oncoplot
    - maftools
    - R_plot
---

# 前言
- 基于maftools object与complexheatmap的oncoplot绘制

# 代码
```r 
# the function is for oncoplot with oncoprint
# author: laojp
# time: 2024.03.01
# position: SYSUCC bioinformatic platform
# parameter: maftools_object, clinical_annotation, annotation_to_show, gene_exclude
# update needed: 
#  1. convert dataframe to maftools object
#  2. anno_simple to show legend
#  3. row annotation update
#  4. annotation to sort
#  5. color freedom

draw_complexheatmap <- function(somatic_maf,group,annotation,gene_exclude=NULL){
  
  cat("#------to get the un_empty sample------\n")
  snv_data <- somatic_maf@data %>% 
    dplyr::filter(!(Variant_Classification %in% c("Amp","Del")),
                  !(Hugo_Symbol %in% gene_exclude)) %>%
    dplyr::select(Tumor_Sample_Barcode,Hugo_Symbol,Variant_Classification) %>% distinct() %>%
    dplyr::group_by(Tumor_Sample_Barcode,Hugo_Symbol) %>%
    dplyr::mutate(Variant_Classification=ifelse(length(unique(Variant_Classification))==1,as.character(Variant_Classification),"Multi_Hit")) %>% ungroup() 
  cnv_data <- somatic_maf@data %>% 
    dplyr::filter(Variant_Classification %in% c("Amp","Del"),
                  !(Hugo_Symbol %in% gene_exclude)) %>%
    dplyr::select(Tumor_Sample_Barcode,Hugo_Symbol,Variant_Classification) %>% distinct()
  
  cat("#------merge empty and unempty data-----\n")  
  data_draw <- dplyr::bind_rows(
    snv_data,cnv_data) %>%
    dplyr::group_by(Tumor_Sample_Barcode,Hugo_Symbol) %>%
    dplyr::summarise(Variant_Classification=paste(unique(Variant_Classification),collapse=";")) %>% ungroup() %>%
    tidyr::pivot_wider(names_from = Tumor_Sample_Barcode,values_from = Variant_Classification,values_fill = "") %>%
    tibble::column_to_rownames(var="Hugo_Symbol")
  cat(paste("gene has: ", dim(data_draw)[1],"\n","unempty samples have: ",dim(data_draw)[2],"\n",sep=""))
  fill_sample <- somatic_maf@variant.classification.summary$Tumor_Sample_Barcode[!(somatic_maf@variant.classification.summary$Tumor_Sample_Barcode) %in%
                                                                                   unique(c(snv_data$Tumor_Sample_Barcode,cnv_data$Tumor_Sample_Barcode))]
  fill_matrix <- matrix(c(""),ncol=length(fill_sample),nrow=dim(data_draw)[1],dimnames = list(c(rownames(data_draw)),c(fill_sample)))
  cat(paste("gene has: ", dim(fill_matrix)[1],"\n","empty samples have: ",dim(fill_matrix)[2],"\n",sep=""))
  data_draw <- cbind(data_draw,fill_matrix)
  
  cat("#------sort sample according to annotation, TMB------\n")
  cat("#------sort gene according to sample frequency------\n")
  # ordered by genes with much mutations, sample with annotated and mutations
  sample_order <- somatic_maf@variant.classification.summary %>% as_tibble() %>%
    dplyr::left_join(group,by=c("Tumor_Sample_Barcode")) %>%
    dplyr::arrange(across(annotation),desc(total))
  data_draw <- data_draw[match((somatic_maf@gene.summary %>% dplyr::filter(!(Hugo_Symbol %in% gene_exclude)))$Hugo_Symbol,rownames(data_draw)),
                         match(sample_order$Tumor_Sample_Barcode,colnames(data_draw))]
  
  cat("#------set color according to mutations------\n")
  qual_col_pals=brewer.pal.info[brewer.pal.info$category == "qual",]#最多78种颜色
  col_vector=unlist(mapply(brewer.pal, qual_col_pals$maxcolors,rownames(qual_col_pals)))
  mutation_type <- c(as.character(somatic_maf@data$Variant_Classification) %>% unique(),"Multi_Hit")
  set.seed(10)
  mutation_type_col=sample(x=col_vector,size=length(mutation_type),replace = FALSE)
  names(mutation_type_col) <- mutation_type
  alter_fun <- list(background = local({
    color = "#CCCCCC"
    function(x, y, w, h) {
          grid.rect(x, y, w, h, gp = gpar(fill = color, col = NA))
        }
    }))
  for (i in names(mutation_type_col)) {
    if(i %in% c("Amp","Del")){
      alter_fun[[i]] <- local({ # need to readed in local cause some problems in closet object, https://github.com/jokergoo/ComplexHeatmap/issues/1041
        color=mutation_type_col[i]
        function(x, y, w, h) {
          grid.rect(x, y, w, h - unit(1, "pt") * 5, gp = gpar(fill = color, col = NA))
        }
      })
    }
    else{
      alter_fun[[i]] <- local({
        color=mutation_type_col[i]
        function(x, y, w, h) {
          grid.rect(x, y, w, h, gp = gpar(fill = color, col = NA))
        }
      })
    }
  }
  
  cat("#------obtain tmb data as top annotation------\n")
  TMB_all <- as.matrix(somatic_maf@variant.classification.summary %>% as_tibble %>%
                     tibble::column_to_rownames(var="Tumor_Sample_Barcode") %>%
                     dplyr::select(-total,-Amp,-Del,-CNV_total))
  TMB_all <- TMB_all[match(colnames(data_draw),rownames(TMB_all)),]
  
  cat("#------oncoprint------\n")
  cat("\t#------anno_list------\n")
  anno_list=list()
  anno_list[["TMB"]] = anno_barplot(TMB_all,gp=gpar(fill=mutation_type_col[match(colnames(TMB_all),names(mutation_type_col))],
                                                    col=mutation_type_col[match(colnames(TMB_all),names(mutation_type_col))]),
                                    height = unit(4, "cm"))
  anno_list[["foo1"]] = anno_empty(border = FALSE)
  for (i in annotation) {
    anno_list[[i]] = sample_order[match(colnames(data_draw[1:20,]),sample_order$Tumor_Sample_Barcode),] %>% dplyr::select(all_of(i)) %>% unlist()
  }
  cat("\t#------all------\n")
  set.seed(100)
  oncoPrint(data_draw[1:20,], alter_fun = alter_fun, col = mutation_type_col, pct_side = "right", row_names_side = "left",
            heatmap_legend_param = list(title = "Alternations", at = names(mutation_type_col), labels = names(mutation_type_col)),
            pct_gp = gpar(fontsize = 14),row_order = c(1:dim(data_draw[1:20,])[1]),column_order = c(1:dim(data_draw[1:20,])[2]),
            bottom_annotation = do.call(HeatmapAnnotation,anno_list),
            top_annotation = NULL)
}

```

